### PR TITLE
New command: skuba addon refresh localconfig (bsc#1173055)

### PR DESCRIPTION
**Pull request by jenting**
## Why is this PR needed?

The addons base manifest will be outdated because skuba did not provide a way to update the base addons manifest.
As time goes by, the addon manifest data structure/key might change, it causes the user's kustomize patch is not valid anymore.

We need to provide a command to upgrade existed base addons manifest, also we'll help to migrate legacy addon manifest folder definition. Note that, we'll not keep the content of legacy addon manifest if the admin had changed the default value in it.

. bsc#1173055 & bsc#1172805
. xrefs https://github.com/SUSE/avant-garde/issues/1765

## What does this PR do?

Add `skuba addon refresh localconfig`: it updates the current on-disk local cluster definition folder to the current Kubernetes version, also it supports migrate old addon definition folders (before 4.2.0) to the new ones.

## Anything else a reviewer needs to know?

N/A

## Info for QA

This is info for QA so that they can validate this. This is **mandatory** if this PR fixes a bug.
If this is a new feature, a good description in "What does this PR do" may be enough.

### Related info

An updated [RFC](https://github.com/SUSE/caasp-rfc/pull/65) for this PR.

### Status **BEFORE** applying the patch

We did not have a command to update the local cluster definition folder before, the only way is to create a temporary cluster definition folder and replace the existed local cluster definition folder. For example:
1. we have existed local cluster definition folder called `my-cluster`
2. create a temporary cluster definition folder: `skuba cluster init --control-plane <IP/FQDN> temp-cluster`
3. copy `temp-cluster/addons/<addon>/base/<addon>.yaml` to  `my-cluster/addons/<addon>/base/<addon>.yaml`

### Status **AFTER** applying the patch

We have a new command to update the local cluster definition folder to the current Kubernetes version. For example:
1. we have existed local cluster definition folder called `my-cluster`
2. run `skuba addon refresh localconfig` inside folder `my-cluster`

## Docs

https://github.com/SUSE/doc-caasp/pull/919

# Merge restrictions

(Please do not edit this)

We are in *v4-maintenance phase*, so we will restrict what can be merged to prevent unexpected surprises:

    What can be merged (merge criteria):
        2 approvals:
            1 developer: code is fine
            1 QA: QA is fine
        there is a PR for updating documentation (or a statement that this is not needed)

### CODE REVIEW

File: pkg/skuba/actions/addon/refresh/localconfig.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	"fmt"
+
+	"github.com/pkg/errors"
+	clientset "k8s.io/client-go/kubernetes"
+
+	"github.com/SUSE/skuba/internal/pkg/skuba/addons"
+	"github.com/SUSE/skuba/internal/pkg/skuba/kubeadm"
+)
+
+// Render implements the `skuba cluster update localconfig` command.
```
Comment (mssola) [inline pkg/skuba/actions/addon/refresh/localconfig.go:30]: Nitpicking: the command now is `refresh`, not `update`.

File: pkg/skuba/actions/addon/refresh/localconfig.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+		ControlPlane:   clusterConfiguration.ControlPlaneEndpoint,
+		ClusterName:    clusterConfiguration.ClusterName,
+	}
+	for addonName, addon := range addons.Addons {
+		if err := addon.Write(addonConfiguration); err != nil {
+			return errors.Wrapf(err, "Failed to refresh addon %s manifest", string(addonName))
+		}
+	}
+
+	fmt.Println("Successfully refresh addons base manifests")
```
Comment (mssola) [inline pkg/skuba/actions/addon/refresh/localconfig.go:54]: `refresh` -> `refreshed`

File: pkg/skuba/actions/addon/refresh/localconfig.go (hunk 1 above)
Comment (jenting) [reply pkg/skuba/actions/addon/refresh/localconfig.go]: nice catch!

File: pkg/skuba/actions/addon/refresh/localconfig.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	"fmt"
+
+	"github.com/pkg/errors"
+	clientset "k8s.io/client-go/kubernetes"
+
+	"github.com/SUSE/skuba/internal/pkg/skuba/addons"
+	"github.com/SUSE/skuba/internal/pkg/skuba/kubeadm"
+)
+
+// Render implements the `skuba addon refresh localconfig` command.
```
Comment (c3y1huang) [inline pkg/skuba/actions/addon/refresh/localconfig.go:30]:
```suggestion
// AddonsBaseManifest implements the `skuba addon refresh localconfig` command.
```

File: pkg/skuba/actions/addon/refresh/localconfig.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+	// re-render all addons manifest
+	addonConfiguration := addons.AddonConfiguration{
+		ClusterVersion: currentClusterVersion,
+		ControlPlane:   clusterConfiguration.ControlPlaneEndpoint,
+		ClusterName:    clusterConfiguration.ClusterName,
+	}
+	for addonName, addon := range addons.Addons {
+		if err := addon.Write(addonConfiguration); err != nil {
+			return errors.Wrapf(err, "Failed to refresh addon %s manifest", string(addonName))
```
Comment (c3y1huang) [inline pkg/skuba/actions/addon/refresh/localconfig.go:50]:
```suggestion
			return errors.Wrapf(err, "failed to refresh addon %s manifest", string(addonName))
```

File: internal/pkg/skuba/addons/addons.go [hunk 5]
```diff
@@ -318,6 +322,15 @@ func (addon Addon) Write(addonConfiguration AddonConfiguration) error {
 	if err := os.MkdirAll(patchResourcesDir, 0700); err != nil {
 		return errors.Wrapf(err, "unable to create directory: %s", patchResourcesDir)
 	}
+
+	// migrates legacy addon manifest if existed
+	lagacyManifestPath := addon.legacyManifestPath(addon.addonDir())
```
Comment (c3y1huang) [inline internal/pkg/skuba/addons/addons.go:327]:
```suggestion
	legacyManifestPath := addon.legacyManifestPath(addon.addonDir())
```

**Comment by innobead** [comment]
LGTM.

**Comment by maximenoel8** [comment]
The feature work correctly. Will need testing when a caasp v4 to v5 is available.

**Pull request merged**